pageserver/src/tenant.rs

```diff
@@ -119,6 +121,234 @@ pub struct Tenant {
+                }
+                Err(e) => {
+                    should_remove_uninit_mark = false;
+                    error!("Failed to clean up uninitialized timeline directory: {e:?}");
+                }
+            }
+
+            if should_remove_uninit_mark {
+                if let Err(e) = uninit_mark.remove_uninit_mark() {
+                    error!("Failed to remove the ininit mark file: {e:?}");
+                }
+            }
+        }
+    }
+}
+
+impl TimelineUninitMark {
+    /// Useful for initializing timelines, existing on disk after the restart.
+    pub fn dummy() -> Self {
+        Self {
+            uninit_mark_deleted: true,
+            uninit_mark_path: PathBuf::new(),
+            timeline_path: PathBuf::new(),
+        }
+    }
+
+    fn new(uninit_mark_path: PathBuf, timeline_path: PathBuf) -> Self {
+        Self {
+            uninit_mark_deleted: false,
+            uninit_mark_path,
+            timeline_path,
+        }
+    }
+
+    fn remove_uninit_mark(mut self) -> anyhow::Result<()> {
+        if !self.uninit_mark_deleted {
+            let uninit_mark_file = &self.uninit_mark_path;
+            let uninit_mark_parent = uninit_mark_file.parent().with_context(|| {
+                format!(
+                    "Uninit mark file {} has no parent",
+                    uninit_mark_file.display()
+                )
+            })?;
+
+            fs::remove_file(&uninit_mark_file)
+                .or_else(|e| {
+                    if e.kind() == io::ErrorKind::NotFound {
+                        Ok(())
+                    } else {
+                        Err(e)
+                    }
+                })
+                .with_context(|| {
+                    format!(
+                        "Failed to remove uninit mark file at path {}",
+                        uninit_mark_file.display()
+                    )
+                })?;
+
+            crashsafe_dir::fsync(uninit_mark_parent)
+                .context("Failed to fsync uninit mark parent")?;
+            self.uninit_mark_deleted = true;
+        }
+
+        Ok(())
+    }
+}
+
+impl Drop for TimelineUninitMark {
+    fn drop(&mut self) {
+        if !self.uninit_mark_deleted {
+            if self.timeline_path.exists() {
+                error!(
+                    "Uninit mark {} is not removed, timeline {} stays uninitialized",
+                    self.uninit_mark_path.display(),
+                    self.timeline_path.display()
+                )
+            } else {
+                // unblock later timeline creation attempts
+                warn!(
+                    "Removing intermediate uninit mark file {}",
+                    self.uninit_mark_path.display()
+                );
+                if let Err(e) = fs::remove_file(&self.uninit_mark_path) {
+                    error!("Failed to remove the uninit mark file: {e}")
+                }
```

Comment:
Similar question about fsync after remove as elsewhere, which can be found on the `TimelineUninitMark::remove_uninit_mark`. I wonder if the same piece of code to delete could be called from both `drop` and `remove_uninit_mark`.

Reply:
Indeed, feels like the most of the deletion could be deduplicated, done.